### PR TITLE
Document Dataset pipeline stage fusion

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

Add docs for the fusion feature (on by default for dataset pipelines).

### CODE REVIEW

File: doc/source/data/key-concepts.rst [hunk 1]
```diff
@@ -137,3 +141,26 @@ Datasets uses the Ray object store to store data blocks, which means it inherits
 **Reference Counting**: Dataset blocks are kept alive by object store reference counting as long as there is any Dataset that references them. To free memory, delete any Python references to the Dataset object.
 
 **Load Balancing**: Datasets uses Ray scheduling hints to spread read tasks out across the cluster to balance memory usage.
+
+Stage Fusion Optimization
+=========================
+
+To avoid unnecessary data movement in the distributed setting, Dataset pipelines will *fuse* compatible stages (i.e., stages with the same compute strategy and resource specifications). Read and map-like stages are always fused if possible. All-to-all dataset transformations such as ``random_shuffle`` can be fused with earlier map-like stages, but not later stages.
```
Comment (jjyao) [inline doc/source/data/key-concepts.rst:148]: Technically, shuffle_reduce can be fused with later map-like stages?

File: doc/source/data/key-concepts.rst [hunk 2]
```diff
@@ -137,3 +141,26 @@ Datasets uses the Ray object store to store data blocks, which means it inherits
...
+
+All of the following optimizations are enabled by default for Dataset pipelines. For Datasets, only read stages are fused. This is since non-pipelined Datasets are eagerly executed except for their initial read stage.
+
+.. code-block:: python
+
+    context = DatasetContext.get_current()
+    context.optimize_fuse_stages = True  # Set to False to disable.
+    context.optimize_fuse_read_stages = True
+    context.optimize_fuse_shuffle_stages = True
```
Comment (jjyao) [inline doc/source/data/key-concepts.rst:157]: Do we want normal users to change those configs directly? Not sure if they should appear in the public doc.

**Review by jjyao — APPROVED** [review]
I'll let @clarkzinzow take a look as well.

File: doc/source/data/key-concepts.rst (hunk 2 above)
Comment (ericl) [reply doc/source/data/key-concepts.rst]: Hmm yeah let me remove this for now.

File: doc/source/data/key-concepts.rst (hunk 1 above)
Comment (ericl) [reply doc/source/data/key-concepts.rst]: It can (@clarkzinzow pointed this out earlier), but it's not yet implemented.

**Pull request merged**